### PR TITLE
gtid, relay: use `gtid_purged` to handle gap in Previous_gtids event

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 

part of #1418 (we need to test master-slave upstream manually later)

### What is changed and how it works?

remove ResetStart, instead, query `@@gtid_purged` and combine it to gtid set when needed

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects
- Potential performance regression

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: relay/relay_test.go [hunk 1]
```diff
@@ -170,14 +171,17 @@ func (t *testRelaySuite) TestTryRecoverLatestFile(c *C) {
 		latestGTIDStr1     = "3ccc475b-2343-11e7-be21-6c0b84d59f30:14"
 		latestGTIDStr2     = "53bfca22-690d-11e7-8a62-18ded7a37b78:495"
 		recoverGTIDSetStr  = "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-17,53bfca22-690d-11e7-8a62-18ded7a37b78:1-505,406a3f61-690d-11e7-87c5-6c92bf46f384:1-456" // 406a3f61-690d-11e7-87c5-6c92bf46f384:123-456 --> 406a3f61-690d-11e7-87c5-6c92bf46f384:1-456
-		greaterGITDSetStr  = "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-20,53bfca22-690d-11e7-8a62-18ded7a37b78:1-510,406a3f61-690d-11e7-87c5-6c92bf46f384:123-456"
+		greaterGITDSetStr  = "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-20,53bfca22-690d-11e7-8a62-18ded7a37b78:1-510,406a3f61-690d-11e7-87c5-6c92bf46f384:1-456"
```
Comment (lance6716) [inline relay/relay_test.go:174]: need to locate why the behaviour of this test changed

File: pkg/utils/db.go [hunk 2]
```diff
@@ -473,3 +473,45 @@ func ExtractTiDBVersion(version string) (*semver.Version, error) {
 	rawVersion = strings.TrimPrefix(rawVersion, "v")
 	return semver.NewVersion(rawVersion)
 }
+
+// AddGSetWithPurged is used to handle this case: https://github.com/pingcap/dm/issues/1418
+// we might get a gtid set from Previous_gtids event in binlog, but that gtid set can't be used to start a gtid sync
+// because it doesn't cover all gtid_purged. The error of using it will be
+// ERROR 1236 (HY000): The slave is connecting using CHANGE MASTER TO MASTER_AUTO_POSITION = 1, but the master has purged binary logs containing GTIDs that the slave requires.
+// so we add gtid_purged to it.
+func AddGSetWithPurged(ctx context.Context, gset gtid.Set, conn *sql.Conn) (gtid.Set, error) {
```
Comment (lichunzhu) [inline pkg/utils/db.go:482]: Why do we use `sql.Conn` instead of `sql.DB` here?

File: relay/relay.go [hunk 3]
```diff
@@ -379,20 +379,17 @@ func (r *Relay) tryRecoverLatestFile(ctx context.Context, parser2 *parser.Parser
...
-					}
+				dbConn, err2 := r.db.Conn(ctx)
+				if err2 != nil {
+					return err2
+				}
+				result.LatestGTIDs, err2 = utils.AddGSetWithPurged(ctx, result.LatestGTIDs, dbConn)
+				if err2 != nil {
+					return err2
+				}
+				latestGTID, err2 = utils.AddGSetWithPurged(ctx, latestGTID, dbConn)
```
Comment (lichunzhu) [inline relay/relay.go:390]: If we don't add purged gtid to `result.LatestGTIDs`, do we need to add purged gtid to `latestGTID` here?

File: pkg/utils/db.go (hunk 2 above)
Comment (lance6716) [reply pkg/utils/db.go:482]: some callers only hold sql.Conn not sql.DB

Comment (lance6716) [reply pkg/utils/db.go:482]: in line 108 of pkg/v1dbschema/schema.go (below file). I'll close connection in next commit

**Review by lichunzhu — COMMENTED** [review]
/lgtm

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**